services/frontend-service/src/ui/Batch.tsx

```diff
@@ -56,9 +68,55 @@ export interface ConfirmationDialogProviderProps {
 const InCart = (actions: BatchAction[], action: BatchAction) =>
     actions ? actions.find((act) => JSON.stringify(act.action) === JSON.stringify(action.action)) : false;
 
+const IsDeployment = (t: ActionTypes) =>
+    t === ActionTypes.Deploy || t === ActionTypes.PrepareUndeploy || t === ActionTypes.Undeploy;
+
+const IsConflictingAction = (cartActions: BatchAction[], newAction: BatchAction) => {
+    for (const action of cartActions) {
+        const act = GetActionDetails(action);
+        const newAct = GetActionDetails(newAction);
+
+        if (IsDeployment(newAct.type) && IsDeployment(act.type)) {
+            if (newAct.application === act.application) {
+                // same app
+                if (newAct.type === ActionTypes.Deploy && newAct.type === act.type) {
+                    // both are deploy actions check env
+                    if (newAct.environment === act.environment) {
+                        // conflict, version doesn't matter
+                        return true;
+                    }
+                } else {
+                    // either one or both are Un-deploying the same app
+                    return true;
+                }
+            }
+        }
+
+        if (newAct.type === ActionTypes.CreateEnvironmentLock && act.type === newAct.type) {
+            if (newAct.environment === act.environment) {
+                // conflict, locking the same env twice
+                return true;
+            }
+        }
+
+        if (newAct.type === ActionTypes.CreateApplicationLock && act.type === newAct.type) {
+            if (newAct.environment === act.environment && newAct.application === act.application) {
+                // conflict, locking the same app/env twice
```

Comment:
same question for application locks

Reply:
Yes. That's what I have in the story. I think it is not useful to lock the same thing twice in one batch at least. you can still lock it in another batch. and you can still apply it anyway.